src/main/java/bio/terra/service/filedata/FileMetadataUtils.java

```diff
@@ -133,4 +136,48 @@ public static List<String> extractDirectoryPaths(String path) {
     }
     return allPaths;
   }
+
+  /**
+   * Results of the successful md5 comparison of user a specified MD5 against a cloud provided MD5
+   *
+   * @param effectiveMd5 The final md5 to record
+   * @param type Which md5 was actually used
+   */
+  public record Md5ValidationResult(String effectiveMd5, Md5Type type) {
```

Comment:
Would it be worth adding a helper instance method to this record to make it easier to figure out whether it's user-provided?  Something like:
```
public boolean isUserProvided() { return type.equals(Md5Type.USER_PROVIDED); }
```

That would make these existing checks a bit easier to read, IMO:
```
finalMd5.type().equals(Md5Type.USER_PROVIDED)
````

Reply:
Hmm good call.  I'll add that